File: aetherment/plugin/Gui/Aeth/TabBar.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using ImGuiNET;

namespace Aetherment.Gui;

// vanilla tabbar ugly af
// it doesnt remember what tab you had like vanilla does
// but its a price im willing to pay

public static partial class Aeth {
	public class TabBar {
		public int Tab;
		public List<string> Tabs;
		
		public TabBar() {
			Tab = 0;
			Tabs = new();
		}
	}
	
	private static Dictionary<uint, TabBar> tabbars = new();
	private static List<TabBar> tabbarstack = new();
	
	public static TabBar BeginTabBar(string sid, bool dockedBottom = true) {
		var id = ImGui.GetID(sid);
		var bar = tabbars.ContainsKey(id) ? tabbars[id] : new TabBar();
		tabbars[id] = bar;
		tabbarstack.Add(bar);
		
		var org = ImGui.GetCursorPos();
		var pos = ImGui.GetCursorScreenPos();
		var barw = ImGui.GetColumnWidth();
		var tabw = barw / Math.Max(1, bar.Tabs.Count);
		var tabh = Aeth.FrameHeight;
		var l = bar.Tabs.Count - 1;
		var hover = -1;
		
		for(int i = 0; i < bar.Tabs.Count; i++) {
			ImGui.SetCursorPos(new Vector2(org.X + i * tabw, org.Y));
			ImGui.Dummy(new Vector2(tabw, tabh));
			if(ImGui.IsItemHovered())
				hover = i;
			if(ImGui.IsItemClicked())
				bar.Tab = i;
		}
		
		var clrt = ImGui.GetColorU32(ImGuiCol.Tab);
		var clrh = ImGui.GetColorU32(ImGuiCol.TabHovered);
		var clra = ImGui.GetColorU32(ImGuiCol.TabActive);
		var draw = ImGui.GetWindowDrawList();
		for(int i = 0; i < bar.Tabs.Count; i++) {
			draw.AddRectFilled(
				new Vector2(pos.X + i * tabw        + (i != 0 ? 1 : 0), pos.Y),
				new Vector2(pos.X + i * tabw + tabw - (i != l ? 1 : 0), pos.Y + tabh),
				i == hover ? clrh : i == bar.Tab ? clra : clrt,
				Aeth.S.TabRounding,
				i == 0 ? (dockedBottom ? ImDrawFlags.RoundCornersTopLeft  : ImDrawFlags.RoundCornersBottomLeft ) :
				i == l ? (dockedBottom ? ImDrawFlags.RoundCornersTopRight : ImDrawFlags.RoundCornersBottomRight) :
				          ImDrawFlags.RoundCornersNone
			);
			
			draw.AddText(
				new Vector2(pos.X + i * tabw, pos.Y) + Aeth.S.FramePadding,
				ImGui.GetColorU32(ImGuiCol.Text),
				bar.Tabs[i]
			);
		}
		
		var y = (dockedBottom ? tabh - 1 : 0);
		draw.AddLine(
			new Vector2(pos.X, pos.Y + y),
			new Vector2(pos.X + barw, pos.Y + y),
			clra
		);
		
		bar.Tabs.Clear();
		
		return bar;
	}
	
	public static void EndTabBar() {
		tabbarstack.RemoveAt(tabbarstack.Count - 1);
	}
	
	public static bool TabItem(string label) {
		var bar = tabbarstack.Last();
		bar.Tabs.Add(label);
		
		return bar.Tab == bar.Tabs.Count - 1;
	}
}

File: aetherment/src/gui/aeth/tabbar.rs
use crate::gui::imgui;
use super::F2 as _;

pub struct TabBar<'a> {
	id: &'a str,
	docked_bottom: bool,
	ignore_add: bool,
	tabs: Vec<&'a str>,
	rect: [f32; 4],
}

impl<'a> TabBar<'a> {
	pub fn dock_top(mut self) -> Self {
		self.docked_bottom = false;
		self
	}
	
	pub fn condition(mut self, condition: bool) -> Self {
		self.ignore_add = !condition;
		self
	}
	
	pub fn tab<F>(mut self, label: &'a str, func: F) -> Self where F: FnOnce() {
		if imgui::get_state_storage().get_i32(imgui::get_id(self.id), 0) == self.tabs.len() as i32 {
			func();
		}
		if !self.ignore_add {
			self.tabs.push(label);
		}
		self.ignore_add = false;
		self
	}
	
	pub fn finish(self) {
		let id = imgui::get_id(self.id);
		let curtab = imgui::get_state_storage().i32(id, 0);
		let s = imgui::get_style();
		let draw = imgui::get_window_draw_list();
		
		let trueorg = imgui::get_cursor_pos();
		let org = [self.rect[0], self.rect[1]];
		let barw = self.rect[2];
		let tabw = barw / self.tabs.len().max(1) as f32;
		let tabh = self.rect[3];
		let tabsize = [tabw, tabh];
		let l = self.tabs.len() - 1;
		
		let clrn = imgui::get_color(imgui::Col::Tab);
		let clrh = imgui::get_color(imgui::Col::TabHovered);
		let clra = imgui::get_color(imgui::Col::TabActive);
		let clrt = imgui::get_color(imgui::Col::Text);

		for (i, tab) in self.tabs.iter().enumerate() {
			let pos = [org[0] + tabw * i as f32, org[1]];
			imgui::set_cursor_pos(pos);
			let pos = pos.add(imgui::get_window_pos());
			imgui::dummy(tabsize);
			if imgui::is_item_clicked(imgui::MouseButton::Left) {
				*curtab = i as i32;
			}
			
			draw.add_rect_filled(
				if i != 0 {pos.add([1.0, 0.0])} else {pos},
				pos.add(if i != l {tabsize.sub([1.0, 0.0])} else {tabsize}),
				if imgui::is_item_hovered() {clrh} else if *curtab == i as i32 {clra} else {clrn},
				s.tab_rounding,
				if i == 0 {if self.docked_bottom {imgui::DrawFlags::RoundCornersTopLeft} else {imgui::DrawFlags::RoundCornersBottomLeft}}
				else if i == l {if self.docked_bottom {imgui::DrawFlags::RoundCornersTopRight} else {imgui::DrawFlags::RoundCornersBottomRight}}
				else {imgui::DrawFlags::RoundCornersNone}
			);
			
			draw.add_text(pos.add(s.frame_padding), clrt, tab);
		}
		
		let p = org.add(imgui::get_window_pos()).add(if self.docked_bottom {[0.0, tabh]} else {[0.0, 0.0]});
		draw.add_line(p, p.add([barw, 0.0]), clra, 1.0);
		
		*curtab = (*curtab).min(self.tabs.len() as i32 - 1);
		imgui::set_cursor_pos(trueorg);
	}
}

pub fn tab_bar(id: &str) -> TabBar {
	let pos = imgui::get_cursor_pos();
	let size = [imgui::get_column_width(-1), super::frame_height()];
	imgui::dummy(size);
	
	TabBar {
		id,
		docked_bottom: true,
		ignore_add: false,
		tabs: Vec::new(),
		rect: [pos[0], pos[1], size[0], size[1]],
	}
}